design/proposals/proxy-support/proxy-support.md

```diff
@@ -0,0 +1,157 @@
+
+When Conjur (or DAP) are run behind a load balancer, users are unable to use CIDR restrictions on Hosts or Layers. Furthermore, Load Balancer IP address appears in Audit logs instead of the true client IP address. This is because Conjur references the last entry in the `X-Forwarded-For` header.
+
+This decision was conscious, and meant to prevent IP spoofing. We have reached a point where customers need a mechanism to add known proxies to Conjur/DAP to identify the true identity of the requesting client. 
+
+This design document is meant to provide a possible solution to the challenge of identifying the client IP while preventing IP spoofing.
+
+- [Current Request Flow](#current-request-flow)
+- [Proposed Solution (Summary)](#proposed-solution-summary)
+- [Proposed Solution (Detailed)](#proposed-solution-detailed)
+  - [Adding Proxies with Policy](#adding-proxies-with-policy)
+  - [Limitations](#limitations)
+    - [Proxies cannot be layered](#proxies-cannot-be-layered)
+- [Rejected Alternatives](#rejected-alternatives)
+  - [Configuration using Environment Variables](#configuration-using-environment-variables)
+
+## Current Request Flow
+
+The following visualization offers a typical experience as a requests moves through a Layer 7 load balancers.
+
+![](/design/diagrams/out/proxy-support-overview/proxy-support-overview.png)
+
+As we can see, each load balancer adds its IP address to the `X-Forwarded-For` header.
+
+With our current implementation:
+```ruby
+# Rack::Request
+def ip
+  if addr = @env['HTTP_X_FORWARDED_FOR']
+    addr.split(',').last.strip
+  else
+    @env['REMOTE_ADDR']
+  end
+end
+```
+the IP address of the last load balancer (`10.2.0.1`) would be used as the request IP.
+
+## Proposed Solution (Summary)
+Users can define the proxies in front of a node(s) using Conjur Policy. This enables Conjur to use the first non-proxy IP as the client IP.
+
+## Proposed Solution (Detailed)
+
+### Adding Proxies with Policy
+
+Given the following setup
+![](/design/diagrams/out/proxy-master-node/proxy-master-node.png)
+
+we can apply the following policy:
+```yaml
+- !policy
+  id: conjur/settings/proxies
+  body:
+    - !host 
+      id: dap-master.mycompany.com
+      annotations:
+        proxy/ip-addresses: ['1.2.3.4']
+```
+
+With the above proxied request, Conjur will:
+1. Check if the `X-Forwarded-For` header has any values. In this example it will have the following string: `10.0.0.1, 1.2.3.4`. 
+2. Perform a Regex match using the host ID in the `conjur/settings/proxies` namespace against the request hostname. In this example, we'd match on `dap-master.mycompany.com`.
+3. Grab the array of IP addresses stored in the `proxy/ip-addresses` annotation (`['1.2.3.4']`).
+4. Starting at the **end** of the list of `X-Forwarded-For` entries, return the first IP address that is not present in the `proxy/ip-addresses` annotation array.
+   1. Skip `1.2.3.4`
+   2. Return `10.0.0.1`
+
```

Comment:
So if i understand it correctly, all the hosts which defined in conjur/settings/proxies policy are permitted proxies addresses? If yes:

1. Can we configure IP instead of hostname ?
2. In step 2, we are fetching the hostname(or the IP) from the TCP packet ?
3. In step 2, if there is no match, we will return a new error of unpermitted proxy?
4. In step 1, If  `X-Forwarded-For` value is empty the behavior will stay as today ?

Reply:
> Can we configure IP instead of hostname ?
Theoretically, yes 😄 

> In step 2, we are fetching the hostname(or the IP) from the TCP packet ?
I was thinking the originating request's URL, but ideally, the most secure approach is best.

> In step 2, if there is no match, we will return a new error of unpermitted proxy?
Yes.

> In step 1, If X-Forwarded-For value is empty the behavior will stay as today ?
Yes.